src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
 			break;
 			}
 
-	// Ideally, we would always call select on the fds to see which
-	// are ready, and return the soonest. Unfortunately, that'd mean
-	// one select-call per packet, which we can't afford in high-volume
-	// environments.  Thus, we call select only every SELECT_FREQUENCY
-	// call (or if all sources report that they are dry).
+	// If there aren't any sources and exit_only_after_terminate is false, just
+	// return an empty set of sources. We want the main loop to end.
+	if ( Size() == 0 && ( ! BifConst::exit_only_after_terminate || terminating ) )
+		return ready;
 
-	++call_count;
+	double timeout = -1;
+	IOSource* timeout_src = nullptr;
+	timer_expired = false;
 
-	IOSource* soonest_src = 0;
-	double soonest_ts = 1e20;
-	double soonest_local_network_time = 1e20;
-	bool all_idle = true;
+	PktSrc* pkt_src = pkt_srcs.empty() ? nullptr : pkt_srcs.front();
 
-	// Find soonest source of those which tell us they have something to
-	// process.
-	for ( SourceList::iterator i = sources.begin(); i != sources.end(); ++i )
+	// If we have an active source of data then find the next timeout value from
+	// those sources. Start by getting the next timeout from the timer manager
+	// since it's not an IOSource.
+	// TODO: how do we handle multiple timer managers here?
```

Comment:
Just an ack that we do need to consider that case. The multiple-timer-mgrs capabilities isn't really used for anything currently, but I'd much prefer not to break it.

Reply:
Maybe what I need is a timer-manager-manager 😁